File: 19. Objective-C/obj-c HW6/obj-c HW6/ViewController.m
//
//  ViewController.m
//  obj-c HW6
//
//  Created by Elena on 12.04.2024.
//

#import "ViewController.h"
#import "Robot.h"

@interface ViewController ()

@property (strong, nonatomic, readonly) UITextField *nameTextField, *xCoordTextField, *yCoordTextField;
@property (strong, nonatomic, readonly) UITextView *textView;
@property (strong, nonatomic, readonly) UIButton *userDefaultsButton, *fileButton;
@property (strong, nonatomic, readonly) NSUserDefaults *userDefaults;
@property (strong, nonatomic, readonly) NSFileManager *fileManager;

@end

@implementation ViewController

#pragma mark - UI components
@synthesize nameTextField = _nameTextField;
- (UITextField *)nameTextField {
    if (!_nameTextField) {
        _nameTextField = [UITextField new];
        _nameTextField.backgroundColor = [UIColor systemTealColor];
        _nameTextField.borderStyle = UITextBorderStyleRoundedRect;
        _nameTextField.placeholder = @"Enter name";
    }
    return _nameTextField;
}

@synthesize xCoordTextField = _xCoordTextField;
- (UITextField *)xCoordTextField {
    if (!_xCoordTextField) {
        _xCoordTextField = [UITextField new];
        _xCoordTextField.backgroundColor = [UIColor systemMintColor];
        _xCoordTextField.borderStyle = UITextBorderStyleRoundedRect;
        _xCoordTextField.keyboardType = UIKeyboardTypeNumberPad;
        _xCoordTextField.placeholder = @"Enter X-coordinate";
    }
    return _xCoordTextField;
}

@synthesize yCoordTextField = _yCoordTextField;
- (UITextField *)yCoordTextField {
    if (!_yCoordTextField) {
        _yCoordTextField = [UITextField new];
        _yCoordTextField.backgroundColor = [UIColor systemMintColor];
        _yCoordTextField.borderStyle = UITextBorderStyleRoundedRect;
        _yCoordTextField.keyboardType = UIKeyboardTypeNumberPad;
        _yCoordTextField.placeholder = @"Enter Y-Coordinate";
    }
    return _yCoordTextField;
}

@synthesize textView = _textView;
- (UITextView *)textView {
    if (!_textView) {
        _textView = [UITextView new];
        _textView.backgroundColor = [UIColor systemGray6Color];
        _textView.layer.cornerRadius = 5;
        _textView.editable = false;
        _textView.font = [UIFont preferredFontForTextStyle:UIFontTextStyleSubheadline];
    }
    return _textView;
}

@synthesize userDefaultsButton = _userDefaultsButton;
- (UIButton *)userDefaultsButton {
    if (!_userDefaultsButton) {
        _userDefaultsButton = [UIButton new];
        [_userDefaultsButton setTitle:@"Save to UserDefaults" forState:UIControlStateNormal];
        _userDefaultsButton.configuration = [UIButtonConfiguration borderedProminentButtonConfiguration];
        [_userDefaultsButton addTarget:self
                                action:@selector(saveToUserDefaultAndReadToTextView)
                      forControlEvents:UIControlEventTouchUpInside];
    }
    return _userDefaultsButton;
}

@synthesize fileButton = _fileButton;
- (UIButton *)fileButton {
    if (!_fileButton) {
        _fileButton = [UIButton new];
        [_fileButton setTitle:@"Save to File" forState:UIControlStateNormal];
        _fileButton.configuration = [UIButtonConfiguration borderedProminentButtonConfiguration];
        [_fileButton addTarget:self
                        action:@selector(saveToFileAndReadToTextView)
              forControlEvents:UIControlEventTouchUpInside];
    }
    return _fileButton;
}

#pragma mark - Lifecycle
- (instancetype)init {
    self = [super init];
    if (self) {
        _userDefaults = [NSUserDefaults standardUserDefaults];
        _fileManager = [NSFileManager defaultManager];
    }
    return self;
}

- (void)viewDidLoad {
    [super viewDidLoad];
    [self setupUI];
}

#pragma mark - Setup UI
- (void)setupUI {
    self.view.backgroundColor = [UIColor systemBackgroundColor];
    [self addSubviews];
    [self setupConstraints];
}

- (void)addSubviews {
    [self.view addSubview:self.nameTextField];
    [self.view addSubview:self.xCoordTextField];
    [self.view addSubview:self.yCoordTextField];
    [self.view addSubview:self.textView];
    [self.view addSubview:self.userDefaultsButton];
    [self.view addSubview:self.fileButton];
}

- (void)setupConstraints {
    self.nameTextField.translatesAutoresizingMaskIntoConstraints = false;
    self.xCoordTextField.translatesAutoresizingMaskIntoConstraints = false;
    self.yCoordTextField.translatesAutoresizingMaskIntoConstraints = false;
    self.textView.translatesAutoresizingMaskIntoConstraints = false;
    self.userDefaultsButton.translatesAutoresizingMaskIntoConstraints = false;
    self.fileButton.translatesAutoresizingMaskIntoConstraints = false;
    
    [NSLayoutConstraint activateConstraints:@[
        [self.nameTextField.topAnchor constraintEqualToAnchor:self.view.safeAreaLayoutGuide.topAnchor constant:25],
        [self.nameTextField.leadingAnchor constraintEqualToAnchor:self.view.safeAreaLayoutGuide.leadingAnchor constant:10],
        [self.nameTextField.trailingAnchor constraintEqualToAnchor:self.view.safeAreaLayoutGuide.trailingAnchor constant:-10],
        
        [self.xCoordTextField.topAnchor constraintEqualToAnchor:self.nameTextField.bottomAnchor constant:10],
        [self.xCoordTextField.leadingAnchor constraintEqualToAnchor:self.nameTextField.leadingAnchor],
        [self.xCoordTextField.widthAnchor constraintEqualToConstant:self.view.frame.size.width / 2 - 15],
        
        [self.yCoordTextField.topAnchor constraintEqualToAnchor:self.xCoordTextField.topAnchor],
        [self.yCoordTextField.trailingAnchor constraintEqualToAnchor:self.nameTextField.trailingAnchor],
        [self.yCoordTextField.widthAnchor constraintEqualToAnchor:self.xCoordTextField.widthAnchor],
        
        [self.textView.topAnchor constraintEqualToAnchor:self.xCoordTextField.bottomAnchor constant:10],
        [self.textView.leadingAnchor constraintEqualToAnchor:self.nameTextField.leadingAnchor],
        [self.textView.trailingAnchor constraintEqualToAnchor:self.nameTextField.trailingAnchor],
        
        [self.userDefaultsButton.topAnchor constraintEqualToAnchor:_textView.bottomAnchor constant:10],
        [self.userDefaultsButton.leadingAnchor constraintEqualToAnchor:self.nameTextField.leadingAnchor],
        [self.userDefaultsButton.trailingAnchor constraintEqualToAnchor:self.nameTextField.trailingAnchor],
        
        [self.fileButton.topAnchor constraintEqualToAnchor:self.userDefaultsButton.bottomAnchor constant:10],
        [self.fileButton.leadingAnchor constraintEqualToAnchor:self.nameTextField.leadingAnchor],
        [self.fileButton.trailingAnchor constraintEqualToAnchor:self.nameTextField.trailingAnchor],
        [self.fileButton.bottomAnchor constraintEqualToAnchor:self.view.safeAreaLayoutGuide.bottomAnchor constant:-25]
    ]];
}

#pragma mark - UserDefaults
- (void)saveToUserDefaults:(Robot *) robot {
    NSData *robotData = [NSKeyedArchiver archivedDataWithRootObject:robot
                                              requiringSecureCoding:YES
                                                              error:nil];
    [self.userDefaults setObject:robotData forKey:@"robot"];
}

- (Robot *)getFromUserDefaults {
    NSData *robotData = [self.userDefaults objectForKey:@"robot"];
    return [NSKeyedUnarchiver unarchivedObjectOfClasses:[NSSet setWithObjects:
                                                         Robot.class,
                                                         NSString.class,
                                                         nil]
                                               fromData:robotData error:nil];
}

- (void)saveToUserDefaultAndReadToTextView {
    Robot *robot = [[Robot alloc]initWithName:self.nameTextField.text
                                  xCoordinate:[self.xCoordTextField.text doubleValue]
                                  yCoordinate:[self.yCoordTextField.text doubleValue]];
    [self saveToUserDefaults:robot];
    Robot *robotFromUserDefaults = [self getFromUserDefaults];
    self.textView.text = [NSString stringWithFormat:@"%@\nX-coordinate: %f\nY-coordinate: %f",
                          robotFromUserDefaults.name,
                          robotFromUserDefaults.xCoordinate,
                          robotFromUserDefaults.yCoordinate];
}

#pragma mark - Saving to file
- (void)saveToFile:(Robot *) robot {
    NSURL *url = [self.fileManager URLsForDirectory:NSDocumentDirectory inDomains:NSUserDomainMask].firstObject;
    NSURL *folderURL = [url URLByAppendingPathComponent:@"Robots"];
    NSURL *fileURL = [folderURL URLByAppendingPathComponent:robot.name];
    NSData *robotData = [NSKeyedArchiver archivedDataWithRootObject:robot
                                              requiringSecureCoding:YES
                                                              error:nil];
    if ([self.fileManager createDirectoryAtURL:folderURL
                   withIntermediateDirectories:YES
                                    attributes:nil
                                         error:nil]) {
        [self.fileManager createFileAtPath:fileURL.path contents:robotData attributes:nil];
    }
}

- (Robot *)getFromFile {
    NSURL *url = [self.fileManager URLsForDirectory:NSDocumentDirectory inDomains:NSUserDomainMask].firstObject;
    NSURL *folderURL = [url URLByAppendingPathComponent:@"Robots"];
    NSURL *fileURL = [folderURL URLByAppendingPathComponent:self.nameTextField.text];
    NSData *robotData;
    if ([self.fileManager fileExistsAtPath:fileURL.path]) {
        robotData = [self.fileManager contentsAtPath:fileURL.path];
    }
    Robot *robot = [NSKeyedUnarchiver unarchivedObjectOfClasses:[NSSet setWithObjects:
                                                                 Robot.class,
                                                                 NSString.class,
                                                                 nil]
                                                       fromData:robotData error:nil];
    return robot;
}

- (void)saveToFileAndReadToTextView {
    Robot *robot = [[Robot alloc]initWithName:self.nameTextField.text
                                  xCoordinate:[self.xCoordTextField.text doubleValue]
                                  yCoordinate:[self.yCoordTextField.text doubleValue]];
    [self saveToFile:robot];
    Robot *robotFromFile = [self getFromFile];
    self.textView.text = [NSString stringWithFormat:@"%@\nX-coordinate: %f\nY-coordinate: %f",
                          robotFromFile.name,
                          robotFromFile.xCoordinate,
                          robotFromFile.yCoordinate];
}

@end

File: 18. Swift - App/Sem1/1-Sem_HW/GB.CreateApp.Homework1/GB.CreateApp.Homework1/ViewController.swift
//
//  ViewController.swift
//  GB.CreateApp.Homework1
//
//  Created by MacOSX on 11.03.2024.
//

import UIKit

class ViewController: UIViewController {
    
    private var imageView = UIImageView(image: UIImage(named: "logo"))
    
    private var label: UILabel = {
        let label = UILabel()
        label.text = "Авторизация"
        label.textAlignment = .center
        label.font = UIFont(name:"Papyrus", size: 24)
        label.textColor = .black
        return label
    }()
    
    private var loginField: UITextField = {
        let login = UITextField()
        login.borderStyle = .line
        login.layer.borderWidth = 1
        login.layer.borderColor = UIColor.black.cgColor
        login.textColor = .black
        let placeholderText = NSAttributedString(string: "Логин",
                                                 attributes: [NSAttributedString.Key.foregroundColor: UIColor.gray])
        login.attributedPlaceholder = placeholderText
        return login
    }()
    
    private var passwordField: UITextField = {
        let password = UITextField()
        password.borderStyle = .line
        password.layer.borderWidth = 1
        password.layer.borderColor = UIColor.black.cgColor
        password.textColor = .black
        let placeholderText = NSAttributedString(string: "Пароль",
                                                    attributes: [NSAttributedString.Key.foregroundColor: UIColor.gray])
        password.attributedPlaceholder = placeholderText
        return password
    }()
    
    private var enterButton: UIButton = {
        let button = UIButton()
        button.setTitle("Войти", for: .normal)
        button.setTitleColor(.black, for: .normal)
        return button
    }()

    override func viewDidLoad() {
            super.viewDidLoad()
            view.backgroundColor = .white
            setupViews()
        }
        
        private func setupViews() {
            view.addSubview(imageView)
            view.addSubview(label)
            view.addSubview(loginField)
            view.addSubview(passwordField)
            view.addSubview(enterButton)
            setupConstraints()
        }

        private func setupConstraints() {
            imageView.translatesAutoresizingMaskIntoConstraints = false
            label.translatesAutoresizingMaskIntoConstraints = false
            loginField.translatesAutoresizingMaskIntoConstraints = false
            passwordField.translatesAutoresizingMaskIntoConstraints = false
            enterButton.translatesAutoresizingMaskIntoConstraints = false
            
            NSLayoutConstraint.activate([
                imageView.topAnchor.constraint(equalTo: view.safeAreaLayoutGuide.topAnchor, constant: 20),
                imageView.centerXAnchor.constraint(equalTo: view.centerXAnchor),
                imageView.widthAnchor.constraint(equalToConstant: view.frame.size.width/4),
                imageView.heightAnchor.constraint(equalToConstant: view.frame.size.width/3.5),
                
                label.topAnchor.constraint(equalTo: imageView.bottomAnchor, constant: 20),
                label.leadingAnchor.constraint(equalTo: view.leadingAnchor, constant: 10),
                label.trailingAnchor.constraint(equalTo: view.trailingAnchor, constant: -10),
                label.heightAnchor.constraint(equalToConstant: 100),
                
                loginField.topAnchor.constraint(equalTo: label.bottomAnchor, constant: 30),
                loginField.leadingAnchor.constraint(equalTo: view.leadingAnchor, constant: 30),
                loginField.trailingAnchor.constraint(equalTo: view.trailingAnchor, constant: -30),
                loginField.heightAnchor.constraint(equalToConstant: 50),
                
                passwordField.topAnchor.constraint(equalTo: loginField.bottomAnchor, constant: 10),
                passwordField.leadingAnchor.constraint(equalTo: view.leadingAnchor, constant: 30),
                passwordField.trailingAnchor.constraint(equalTo: view.trailingAnchor, constant: -30),
                passwordField.heightAnchor.constraint(equalToConstant: 50),
                
                enterButton.topAnchor.constraint(equalTo: passwordField.bottomAnchor, constant: 30),
                enterButton.centerXAnchor.constraint(equalTo: view.centerXAnchor),
                enterButton.widthAnchor.constraint(equalToConstant: view.frame.size.width/4),
                enterButton.heightAnchor.constraint(equalToConstant: view.frame.size.width/4)
            ])
        }

    }

